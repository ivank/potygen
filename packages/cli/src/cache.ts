import { existsSync, readFileSync, writeFileSync, statSync, mkdirSync } from 'fs';
import { dirname } from 'path';
import { createHash } from 'crypto';

export class CachedFileParser<TResult> {
  constructor(public store = new Map<string, TResult>(), public process: (content: string, path: string) => TResult) {}

  public get(path: string): TResult {
    const content = readFileSync(path, 'utf-8');
    const key = createHash('md5').update(content).digest('hex');
    const cachedResult = this.store.get(key);
    if (cachedResult) {
      return cachedResult;
    } else {
      const result = this.process(content, path);
      this.store.set(key, result);
      return result;
    }
  }
}

export class CachedProcessParser<TResult> {
  constructor(public store = new Map<string, TResult>(), public process: (content: string, path: string) => TResult) {}

  public get(path: string): TResult {
    const content = readFileSync(path, 'utf-8');
    const key = createHash('md5').update(content).digest('hex');
    const cachedResult = this.store.get(key);
    if (cachedResult) {
      return cachedResult;
    } else {
      const result = this.process(content, path);
      this.store.set(key, result);
      return result;
    }
  }
}

export class CacheStore {
  public store = new Map<string, number>();

  constructor(public fileName: string, public enabled: boolean = false, public cacheClear: boolean = false) {
    if (this.enabled && !this.cacheClear) {
      if (existsSync(fileName)) {
        this.store = new Map(Object.entries(JSON.parse(readFileSync(fileName, 'utf-8'))));
      }
    }
  }

  public shouldParseFile(path: string) {
    if (!this.enabled) {
      return true;
    } else {
      const mtime = statSync(path).mtime.getTime();
      const currentMtime = this.store.get(path);
      if (currentMtime && mtime === currentMtime) {
        return false;
      }
    }
    return true;
  }

  public cacheFile(path: string) {
    const mtime = statSync(path).mtime.getTime();
    this.store.set(path, mtime);
  }

  public save() {
    if (this.enabled) {
      mkdirSync(dirname(this.fileName), { recursive: true });
      writeFileSync(this.fileName, JSON.stringify(Object.fromEntries(this.store)), 'utf-8');
    }
  }
}
