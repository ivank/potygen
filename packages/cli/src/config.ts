import { Record, Optional, Static, String, Boolean } from 'runtypes';

export const Config = Record({
  files: Optional(String),
  watch: Optional(Boolean),
  root: Optional(String),
  connection: Optional(String),
  template: Optional(String),
  verbose: Optional(Boolean),
  silent: Optional(Boolean),
  typePrefix: Optional(String),
  preload: Optional(Boolean),
  cache: Optional(Boolean),
  cacheFile: Optional(String),
  cacheClear: Optional(Boolean),
});

export type ConfigType = Static<typeof Config>;
export type FullConfigType = Required<ConfigType>;

export const toConfig = (config: unknown): FullConfigType => ({
  files: '**/*.sql',
  cacheFile: '.cache/potygen.cache',
  root: process.cwd(),
  preload: false,
  template: '{{dir}}/{{name}}.queries.ts',
  connection: 'postgres://localhost:5432/db',
  watch: false,
  verbose: false,
  silent: false,
  cache: false,
  cacheClear: false,
  typePrefix: '',
  ...Config.check(config),
});
