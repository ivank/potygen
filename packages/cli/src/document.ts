import { document, printDocument, DocumentContext, Type, withIdentifier } from '@ovotech/ts-compose';
import { TypeNode } from 'typescript';
import { isLoadedArrayType, isLoadedUnionType, isLoadedLiteralType, isLoadedValuesPick } from './guards';
import { LoadedQuery, LoadedValuesPick, LoadedType } from './types';

const toPropertyType = (type: LoadedType | LoadedValuesPick): TypeNode => {
  if (isLoadedValuesPick(type)) {
    return Type.Array(
      Type.TypeLiteral({
        props: type.items.map((item) => Type.Prop({ name: item.name, type: toPropertyType(item.value) })),
      }),
    );
  } else if (isLoadedArrayType(type)) {
    return Type.Array(toPropertyType(type.items));
  } else if (isLoadedUnionType(type)) {
    return Type.Union(type.items.map(toPropertyType));
  } else if (isLoadedLiteralType(type)) {
    return Type.Literal(type.value);
  } else {
    switch (type.type) {
      case 'Date':
        return Type.Referance('Date');
      case 'boolean':
        return Type.Boolean;
      case 'json':
        return Type.Unknown;
      case 'null':
        return Type.Null;
      case 'number':
        return Type.Number;
      case 'string':
        return Type.String;
      case 'unknown':
        return Type.Unknown;
    }
  }
};

export const toQueryTypescriptTypes = (context: DocumentContext, query: LoadedQuery) => {
  const paramsContext = withIdentifier(
    context,
    Type.Interface({
      name: 'QueryParams',
      isExport: true,
      props: query.params.map(({ name, type }) =>
        Type.Prop({ name, type: toPropertyType(type), isOptional: type.optional }),
      ),
    }),
  );

  const resultContext = withIdentifier(
    paramsContext,
    Type.Interface({
      name: 'QueryResult',
      isExport: true,
      props: query.result.map(({ name, type }) =>
        Type.Prop({ name, type: toPropertyType(type), isOptional: type.optional }),
      ),
    }),
  );

  return document(
    resultContext,
    Type.Interface({
      name: 'Query',
      isExport: true,
      props: [
        Type.Prop({ name: 'params', type: Type.Referance('QueryParams') }),
        Type.Prop({ name: 'result', type: Type.Referance('QueryResult') }),
      ],
    }),
  );
};

export const toQueryTypescript = (loadedQuery: LoadedQuery): string => {
  return printDocument(toQueryTypescriptTypes({}, loadedQuery));
};
