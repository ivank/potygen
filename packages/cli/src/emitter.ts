import { existsSync } from 'fs';
import {
  createSourceFile,
  ScriptTarget,
  isTaggedTemplateExpression,
  Node,
  isImportDeclaration,
  isStringLiteral,
  isNoSubstitutionTemplateLiteral,
  isNamedImports,
  SourceFile,
  isIdentifier,
  isCallExpression,
} from 'typescript';
import { basename, dirname, relative } from 'path';
import {
  parser,
  toQueryInterface,
  toLoadedQueryInterface,
  LoadError,
  ParsedSqlFileLoadError,
  ParsedTypescriptFileLoadError,
  ParseError,
  LoadContext,
  LoadedData,
  LoadedFile,
  ParsedFile,
  TemplateTagQuery,
  loadAllData,
  toMilliseconds,
} from '@potygen/potygen';
import { toTypeScriptPrinter } from './typescript-printer';
import { mkdir, readFile, stat, writeFile } from 'fs/promises';
import { createHash } from 'crypto';

/**
 * Handle both "last updated" checks on files as well as actually checking file contents.
 */
export class CacheStore {
  public contents = new Map<string, { path: string; content: string } | undefined>();
  public updates = new Map<string, number>();

  constructor(
    /**
     * Path to the cache file
     */
    public fileName: string,
    /**
     * If enabled is false, do not perform any caching
     */
    public enabled: boolean = false,
    /**
     * If true, initialize the cache with empty values so that after save everything is overwritten
     */
    public cacheClear: boolean = false,
    /**
     * Current version of the cache, if version mismatch, clear the cache
     */
    public cacheVersion = '1.0',
  ) {}

  /**
   * Check the file's last modified time (mtime).
   * If its after the time stored in cache, consider the file stale.
   */
  public async isStale(path: string): Promise<boolean> {
    if (!this.enabled) {
      return true;
    }
    const mtime = (await stat(path)).mtime.getTime();
    const currentMtime = this.updates.get(path);
    return !(currentMtime && mtime === currentMtime);
  }

  /**
   * For a given path and file content, check if content is present in the cache.
   * If it is not, run the process function and store the results in the cache,
   * keyed by md5 of the initial content
   */
  public async cachedOrProcessed(
    path: string,
    content: string,
    process: (path: string, content: string) => Promise<{ path: string; content: string } | undefined>,
  ): Promise<{ path: string; content: string; isCached?: boolean } | undefined> {
    if (!this.enabled) {
      return await process(path, content);
    }
    const key = createHash('md5').update(content).digest('hex');
    const cached = this.contents.get(key);
    const mtime = (await stat(path)).mtime.getTime();
    if (cached) {
      this.updates.set(path, mtime);
      return { ...cached, isCached: true };
    } else {
      const processed = await process(path, content);
      const mtime = (await stat(path)).mtime.getTime();
      this.contents.set(key, processed);
      this.updates.set(path, mtime);
      return processed;
    }
  }

  /**
   * Load the cache from file
   */
  public async load(): Promise<void> {
    if (!this.cacheClear && existsSync(this.fileName)) {
      const cache = JSON.parse(await readFile(this.fileName, 'utf-8'));
      if (typeof cache === 'object' && cache !== null && 'version' in cache && cache.version === this.cacheVersion) {
        this.updates = new Map(Object.entries(cache.updates));
        this.contents = new Map(Object.entries(cache.contents));
      }
    }
  }

  /**
   * Save current state in the cache file
   */
  public async save() {
    await mkdir(dirname(this.fileName), { recursive: true });
    await writeFile(
      this.fileName,
      JSON.stringify({
        updates: Object.fromEntries(this.updates),
        contents: Object.fromEntries(this.contents),
        version: this.cacheVersion,
      }),
      'utf-8',
    );
  }
}

const toTemplateParent = (node: Node): Node =>
  isCallExpression(node.parent) ? toTemplateParent(node.parent) : node.parent;

/**
 * Extract the sql tagged template literals from a typescript ast
 */
const getTemplateTagQueries = (ast: SourceFile): TemplateTagQuery[] => {
  const queries: TemplateTagQuery[] = [];
  let tagPropertyName = 'sql';

  const visitor = (node: Node): void => {
    if (
      isImportDeclaration(node) &&
      isStringLiteral(node.moduleSpecifier) &&
      node.importClause?.namedBindings &&
      isNamedImports(node.importClause.namedBindings) &&
      node.moduleSpecifier.text === '@potygen/potygen'
    ) {
      tagPropertyName =
        node.importClause?.namedBindings.elements.find(
          ({ propertyName, name }) => (propertyName ?? name).text === 'sql',
        )?.name.text ?? tagPropertyName;
    } else if (
      isTaggedTemplateExpression(node) &&
      isNoSubstitutionTemplateLiteral(node.template) &&
      isIdentifier(node.tag) &&
      node.tag.text === tagPropertyName
    ) {
      const tag = {
        name: toTemplateParent(node).getChildAt(0).getText(),
        pos: node.template.pos + 1,
        template: node.template.text,
      };
      try {
        queries.push({ ...tag, queryInterface: toQueryInterface(parser(node.template.text).ast) });
      } catch (error) {
        throw new ParseError(tag, `Error parsing sql: ${String(error)}`);
      }
    } else {
      node.forEachChild(visitor);
    }
  };

  visitor(ast);
  return queries;
};

const parseFile = (path: string, content: string): ParsedFile | undefined => {
  if (path.endsWith('.ts')) {
    const source = createSourceFile(basename(path), content, ScriptTarget.ES2021, true);
    const file: ParsedFile = { type: 'ts', source, path, queries: getTemplateTagQueries(source) };
    return file.queries.length > 0 ? file : undefined;
  } else {
    return { type: 'sql', path, content, queryInterface: toQueryInterface(parser(content).ast) };
  }
};

const isError = (error: unknown): error is LoadError | ParseError =>
  error instanceof LoadError || error instanceof ParseError;

const toLoadFile =
  (data: LoadedData[]) =>
  (file: ParsedFile): LoadedFile => {
    if (file.type === 'sql') {
      try {
        return { ...file, loadedQuery: toLoadedQueryInterface(data)(file.queryInterface) };
      } catch (error) {
        throw isError(error) ? new ParsedSqlFileLoadError(file, error) : error;
      }
    } else {
      return {
        ...file,
        queries: file.queries.map((template) => {
          try {
            return { ...template, loadedQuery: toLoadedQueryInterface(data)(template.queryInterface) };
          } catch (error) {
            throw isError(error) ? new ParsedTypescriptFileLoadError(file, template, error) : error;
          }
        }),
      };
    }
  };

export const toEmitter = async (
  ctx: LoadContext,
  cacheStore: CacheStore,
  options: { root: string; template: string; typePrefix?: string },
) => {
  const data = await loadAllData(ctx, []);
  const loadFile = toLoadFile(data);
  const typeScriptPrinter = toTypeScriptPrinter(options.root, options.template, options.typePrefix);

  return async (path: string): Promise<{ path: string } | undefined> => {
    const start = process.hrtime();
    const relPath = relative(options.root ?? '.', path);
    try {
      if (await cacheStore.isStale(path)) {
        const content = await readFile(path, 'utf-8');
        const output = await cacheStore.cachedOrProcessed(path, content, async (path, content) => {
          const file = parseFile(path, content);
          return file ? await typeScriptPrinter(loadFile(file)) : undefined;
        });
        if (output) {
          await mkdir(dirname(output.path), { recursive: true });
          await writeFile(output.path, output.content, 'utf-8');
          const elapsed = toMilliseconds(process.hrtime(start));
          ctx.logger.info(`[${output.isCached ? 'Cached' : 'Generated'}]: ${relPath} (${elapsed}ms)`);
        }
      } else {
        ctx.logger.info(`[Not modified]: ${relPath}`);
      }
      return { path };
    } catch (error) {
      const elapsed = toMilliseconds(process.hrtime(start));
      ctx.logger.error(`[Error]: ${relPath} (${elapsed})ms`);
      throw error;
    }
  };
};
