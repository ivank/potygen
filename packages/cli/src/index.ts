export {
  DataTable,
  DataFunction,
  DataEnum,
  DataView,
  DataComposite,
  LoadedDataTable,
  LoadedDataComposite,
  DataViewRaw,
  DataViewParsed,
  LoadedDataView,
  LoadedDataFunction,
  LoadedDataEnum,
  Data,
  LoadedData,
  LoadedParam,
  LoadedResult,
  LoadedQueryInterface,
  LoadedFunction,
  LoadedComposite,
  LoadedSourceTable,
  LoadedSourceQuery,
  LoadedSourceView,
  LoadedSourceValues,
  LoadedSource,
  LoadedContext,
  TemplateTagQuery,
  ParsedTypescriptFile,
  ParsedSqlFile,
  ParsedFile,
  LoadedTemplateTagQuery,
  LoadedTypescriptFile,
  LoadedSqlFile,
  LoadedFile,
  Logger,
  LoadContext,
  QualifiedName,
} from './types';

export {
  isDataTable,
  isDataFunction,
  isDataEnum,
  isLoadedDataTable,
  isLoadedDataFunction,
  isLoadedDataEnum,
  isLoadedDataView,
  isLoadedDataComposite,
} from './guards';

export { toLoadedContext, loadQueryInterfacesData, toLoadedQueryInterface, loadData, extractDataSources } from './load';

export { Config, ConfigType, FullConfigType, toConfig } from './config';

export { emitLoadedFile, toTypeSource } from './emit';
export { glob } from './glob';
