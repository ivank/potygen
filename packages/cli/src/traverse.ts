import { readFileSync, watchFile } from 'fs';
import { Readable, Writable } from 'stream';
import { glob } from './glob';
import {
  createSourceFile,
  ScriptTarget,
  isTaggedTemplateExpression,
  Node,
  isImportDeclaration,
  isStringLiteral,
  isNoSubstitutionTemplateLiteral,
  isNamedImports,
  SourceFile,
  isIdentifier,
  isCallExpression,
} from 'typescript';
import { basename, relative } from 'path';
import {
  parser,
  QueryInterface,
  toQueryInterface,
  loadQueryInterfacesData,
  toLoadedQueryInterface,
  LoadError,
  ParsedSqlFileLoadError,
  ParsedTypescriptFileLoadError,
  ParseError,
  LoadContext,
  LoadedData,
  LoadedFile,
  Logger,
  ParsedFile,
  ParsedSqlFile,
  ParsedTypescriptFile,
  TemplateTagQuery,
  loadAllData,
} from '@potygen/potygen';
import { ClientBase } from 'pg';
import { emitLoadedFile } from './emit';
import { inspect } from 'util';

const toTemplateParent = (node: Node): Node =>
  isCallExpression(node.parent) ? toTemplateParent(node.parent) : node.parent;

const getTemplateTagQueries = (ast: SourceFile): TemplateTagQuery[] => {
  const queries: TemplateTagQuery[] = [];
  let tagPropertyName = 'sql';

  const visitor = (node: Node): void => {
    if (
      isImportDeclaration(node) &&
      isStringLiteral(node.moduleSpecifier) &&
      node.importClause?.namedBindings &&
      isNamedImports(node.importClause.namedBindings) &&
      node.moduleSpecifier.text === '@potygen/potygen'
    ) {
      tagPropertyName =
        node.importClause?.namedBindings.elements.find(
          ({ propertyName, name }) => (propertyName ?? name).text === 'sql',
        )?.name.text ?? tagPropertyName;
    } else if (
      isTaggedTemplateExpression(node) &&
      isNoSubstitutionTemplateLiteral(node.template) &&
      isIdentifier(node.tag) &&
      node.tag.text === tagPropertyName
    ) {
      const tag = {
        name: toTemplateParent(node).getChildAt(0).getText(),
        pos: node.template.pos + 1,
        template: node.template.text,
      };
      try {
        queries.push({ ...tag, queryInterface: toQueryInterface(parser(node.template.text).ast) });
      } catch (error) {
        throw new ParseError(tag, `Error parsing sql: ${error instanceof Error ? error.message : String(error)}`);
      }
    } else {
      node.forEachChild(visitor);
    }
  };

  visitor(ast);
  return queries;
};

const toParsedTypescriptFile = (path: string): ParsedTypescriptFile => {
  const sourceText = readFileSync(path, 'utf-8');
  const source = createSourceFile(basename(path), sourceText, ScriptTarget.ES2021, true);
  return { type: 'ts', source, path, queries: getTemplateTagQueries(source) };
};

const toParsedSqlFile = (path: string): ParsedSqlFile | undefined => {
  const content = readFileSync(path, 'utf-8');
  return { type: 'sql', path, content, queryInterface: toQueryInterface(parser(content).ast) };
};

const toQueryInterfaces = (files: ParsedFile[]): QueryInterface[] =>
  files.flatMap((file) =>
    file.type === 'ts' ? file.queries.map((query) => query.queryInterface) : file.queryInterface,
  );

const loadDataFromParsedFiles = async (
  ctx: LoadContext,
  data: LoadedData[],
  files: ParsedFile[],
): Promise<LoadedData[]> => loadQueryInterfacesData(ctx, toQueryInterfaces(files), data);

const isError = (error: unknown): error is LoadError | ParseError =>
  error instanceof LoadError || error instanceof ParseError;

const loadFile =
  (data: LoadedData[]) =>
  (file: ParsedFile): LoadedFile => {
    if (file.type === 'sql') {
      try {
        return { ...file, loadedQuery: toLoadedQueryInterface(data)(file.queryInterface) };
      } catch (error) {
        throw isError(error) ? new ParsedSqlFileLoadError(file, error) : error;
      }
    } else {
      return {
        ...file,
        queries: file.queries.map((template) => {
          try {
            return { ...template, loadedQuery: toLoadedQueryInterface(data)(template.queryInterface) };
          } catch (error) {
            throw isError(error) ? new ParsedTypescriptFileLoadError(file, template, error) : error;
          }
        }),
      };
    }
  };

const parseFile = (path: string): ParsedFile | undefined => {
  if (path.endsWith('.ts')) {
    const file = toParsedTypescriptFile(path);
    return file.queries.length > 0 ? file : undefined;
  } else {
    return toParsedSqlFile(path);
  }
};

export class SqlRead extends Readable {
  public source: Generator<string, void, unknown>;
  public watchedFiles = new Set<string>();

  constructor(public options: { path: string; root: string; watch: boolean; logger: Logger }) {
    super({ objectMode: true });
    this.source = glob(options.path, options.root);
  }

  next() {
    let path: IteratorResult<string>;
    while (!(path = this.source.next()).done) {
      const file = parseFile(path.value);
      if (file) {
        return file;
      }
    }
    return undefined;
  }

  watchFile(path: string): () => void {
    return () => {
      const file = parseFile(path);
      if (file) {
        this.options.logger.info(`Processing ${relative(this.options.root ?? '.', path)}`);
        this.push(file);
      }
    };
  }

  _read() {
    const next = this.next();
    if (next) {
      if (this.options.watch && !this.watchedFiles.has(next.path)) {
        this.watchedFiles.add(next.path);
        watchFile(next.path, this.watchFile(next.path));
      }
      this.options.logger.info(`Processing ${relative(this.options.root ?? '.', next.path)}`);
      this.push(next);
    } else if (!this.options.watch) {
      this.options.logger.info(`Done`);
      this.push(null);
    }
  }
}

export class QueryLoader extends Writable {
  public ctx: LoadContext;
  public data: LoadedData[] = [];
  constructor(
    public options: {
      db: ClientBase;
      root: string;
      template: string;
      logger: Logger;
      typePrefix?: string;
      preload?: boolean;
    },
  ) {
    super({ objectMode: true });
    this.ctx = { db: options.db, logger: options.logger ?? console };
  }

  async _writev(
    chunks: Array<{ chunk: ParsedFile; encoding: BufferEncoding }>,
    callback: (error?: Error | null) => void,
  ): Promise<void> {
    try {
      const parsedFiles = chunks.map((file) => file.chunk);
      this.ctx.logger.debug(
        `Parse files: ${inspect(
          parsedFiles.map((file) => `${relative(this.options.root, file.path)} (${file.type})`),
        )}`,
      );
      this.data = this.options.preload
        ? this.data.length === 0
          ? await loadAllData(this.ctx, this.data)
          : this.data
        : await loadDataFromParsedFiles(this.ctx, this.data, parsedFiles);
      await Promise.all(
        parsedFiles
          .map(loadFile(this.data))
          .map(emitLoadedFile(this.options.root, this.options.template, this.options.typePrefix)),
      );
    } catch (error) {
      this.options.logger.error(error instanceof Error ? String(error) : new Error(String(error)));
    }
    callback();
  }

  async _write(
    file: ParsedTypescriptFile,
    encoding: BufferEncoding,
    callback: (error?: Error | null) => void,
  ): Promise<void> {
    try {
      this.ctx.logger.debug(`Parse file: ${relative(this.options.root, file.path)} (${file.type})`);
      this.data = this.options.preload
        ? this.data.length === 0
          ? await loadAllData(this.ctx, this.data)
          : this.data
        : await loadDataFromParsedFiles(this.ctx, this.data, [file]);
      await emitLoadedFile(
        this.options.root,
        this.options.template,
        this.options.typePrefix,
      )(loadFile(this.data)(file));
    } catch (error) {
      this.options.logger.error(error instanceof Error ? String(error) : new Error(String(error)));
    }
    callback();
  }
}
