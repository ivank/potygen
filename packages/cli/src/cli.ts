import { Client } from 'pg';
import { pipeline } from 'stream';
import { promisify } from 'util';
import { SqlRead, QueryLoader } from './traverse';

const asyncPipeline = promisify(pipeline);

const main = async () => {
  const db = new Client({
    database: 'schemes',
    user: 'schemes',
    password: 'dev-pass',
    host: '127.0.0.1',
    port: 5432,
    ssl: false,
  });
  try {
    await db.connect();

    const sqls = new SqlRead('src/**/*.query.ts', '../../../schemes/packages/api');
    const sink = new QueryLoader(
      db,
      '../../../schemes/packages/api',
      '{{root}}/src/queries/__psql-ts__/{{name}}.queries.ts',
    );

    await asyncPipeline(sqls, sink);
  } catch (error) {
    console.log(String(error));
  } finally {
    await db.end();
  }
};

main();
