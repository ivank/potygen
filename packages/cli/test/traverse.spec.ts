import { Client } from 'pg';
import { pipeline } from 'stream';
import { promisify } from 'util';

const asyncPipeline = promisify(pipeline);

import { SqlRead, QueryLoader } from '../src/traverse';

// const cwd = (path: string) => join(__dirname, path);

describe('Traverse', () => {
  it('Should work', async () => {
    const db = new Client({ database: 'sql-ast', user: 'sql-ast', password: 'dev-pass' });
    try {
      await db.connect();

      const sqls = new SqlRead('dir/**/*.ts', __dirname);
      const sink = new QueryLoader(db, __dirname, '{{root}}/__generated__/{{name}}.queries.ts');

      await asyncPipeline(sqls, sink);
    } finally {
      await db.end();
    }
  });
});
