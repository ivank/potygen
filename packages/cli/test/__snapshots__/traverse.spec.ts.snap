// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Traverse Should work: __generated__/account-levelisation.upsert.query.queries.ts 1`] = `
"export interface Params {
    levelisationId: unknown;
    start?: Date;
}
export interface Result {
    id: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/account-levelisations.query.queries.ts 1`] = `
"export interface Params {
    q: \\"\\";
    resolvedPostlev: \\"\\" | boolean;
    state: (\\"Done\\" | \\"Pending\\");
    levelisationId: number;
    ids: number[];
    sortField: \\"totalPayment\\" | \\"vatPayment\\" | \\"generationPayment\\" | \\"exportPayment\\" | \\"generationEnergy\\" | \\"exportEnergy\\" | \\"cfrFitId\\" | \\"state\\" | \\"isAccepted\\" | \\"error\\" | \\"exportType\\";
    sortOrder: \\"DESC\\" | \\"ASC\\";
    limit: number;
    offset: number;
}
export interface Result {
    id: number;
    levelisationId: number;
    accountId: number;
    installationId: number;
    state: \\"Done\\" | \\"Pending\\";
    isAccepted: boolean;
    generationStartReadOn?: Date;
    generationStartReadValue?: string;
    generationEndReadOn?: Date;
    generationEndReadValue?: string;
    generationPercentageSplit?: string;
    generationPayment?: string;
    generationEnergy?: string;
    exportStartReadOn?: Date;
    exportStartReadValue?: string;
    exportEndReadOn?: Date;
    exportEndReadValue?: string;
    exportPercentageSplit?: string;
    exportPayment?: string;
    exportEnergy?: string;
    exportType: \\"Off Grid\\" | \\"Deemed\\" | \\"Metered Export\\" | \\"Off Grid\\" | \\"PPA\\" | \\"Deemed\\" | \\"Metered Export\\" | \\"Off Grid\\" | \\"PPA\\" | \\"Deemed\\" | \\"Metered Export\\" | \\"PPA\\";
    technologyType: \\"H\\" | \\"W\\" | \\"AD\\" | \\"CHP\\" | \\"PV\\" | \\"H\\" | \\"W\\" | \\"AD\\" | \\"CHP\\" | \\"PV\\";
    totalPayment?: string;
    vatPayment?: string;
    generationPeriods?: unknown;
    exportPeriods?: unknown;
    errorCode: unknown;
    errorParams: unknown;
    cfrFitId: string;
    resolvedPostlevId?: number;
    isBacsPaymentsSent?: boolean;
    isChequePaymentsSent?: boolean;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/account-levelisations-bulk-close-accounts.query.queries.ts 1`] = `
"export interface Params {
    ids: number[];
}
export interface Result {
    id: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/account-levelisations-set-bulk-update.query.queries.ts 1`] = `
"export interface Params {
    ids: number[];
}
export interface Result {
    id: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/accounts.list.query.queries.ts 1`] = `
"export interface Params {
    q: \\"\\" | string;
    ids: number[];
}
export interface Result {
    total: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/accounts.queries.ts 1`] = `
"export interface Params {
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/accounts.query.queries.ts 1`] = `
"export interface Params {
    q: \\"\\" | string;
    ids: number[];
    limit: number;
    offset: number;
}
export interface Result {
    id: number;
    state: \\"Active\\" | \\"Pending\\" | \\"Dispute\\" | \\"Closed\\";
    startOn?: Date;
    firstName: string;
    lastName: string;
    email?: string;
    address: string;
    fitIds: string[];
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/bacs.query.queries.ts 1`] = `
"export interface Params {
    quarter?: string;
    perPage: string;
    offset: string;
}
export interface Result {
    fitId: string;
    name?: string;
    amount: string;
    quarter?: string;
    bankDetails: string;
    addressLine1?: string;
    addressLine2?: string;
    addressLine3?: string;
    postcode: string;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/contracts.queries.ts 1`] = `
"export interface Params {
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/delete-accounts.query.queries.ts 1`] = `
"export interface Params {
    type: \\"SEG\\" | \\"FIT\\";
    state: \\"Active\\" | \\"Pending\\" | \\"Dispute\\" | \\"Closed\\";
}
export interface Result {
    id: number;
    sourceSystemId?: number;
    customerId: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/generation-read-data.query.queries.ts 1`] = `
"export interface Params {
    installationId: number;
    exportDateOn: Date;
}
export interface Result {
    dateOn: Date;
    value: number;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/installation-meter-update.query.queries.ts 1`] = `
"export interface Params {
    removalDate: unknown;
    meterId: number;
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/installations.queries.ts 1`] = `
"export interface Params {
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/meter-reads.queries.ts 1`] = `
"export interface Params {
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/process-batch.query.queries.ts 1`] = `
"export interface Params {
    fitIds: string[];
}
export interface Result {
    cfrFiTId?: string;
    email?: string;
    contactName: string;
    company?: string;
    line1?: string;
    line2: string;
    town?: string;
    postcode: string;
    county?: string;
    country: string;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/rates-by-tarrif-ids.query.queries.ts 1`] = `
"export interface Params {
    ids: number[][];
}
export interface Result {
    tariffId: number;
    rate: string;
    startOn: Date;
    tariffCode: string;
    tariffType: \\"Export\\" | \\"Generation\\" | \\"SEG\\";
    endOn?: Date;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/read-create.query.queries.ts 1`] = `
"export interface Params {
    meterId: number;
    reason?: string;
    type: \\"Opening\\" | \\"Closing\\" | \\"Quarterly\\" | \\"Meter Verification\\";
    dateOn: Date;
    submittedAt?: Date;
    value: number;
}
export interface Result {
    id: number;
    value: number;
    meterId: number;
    submittedAt?: Date;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/reads-by-account-id.query.queries.ts 1`] = `
"export interface Params {
    id: number;
    offset: string;
    perPage: string;
}
export interface Result {
    id: number;
    msn: string;
    meterId: number;
    dateOn: Date;
    value: number;
    type: \\"Opening\\" | \\"Closing\\" | \\"Quarterly\\" | \\"Meter Verification\\";
    reason?: string;
    createdAt: Date;
    updatedAt?: Date;
    sourceSystemId?: number;
    mpan: string;
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/tables.queries.ts 1`] = `
"export interface Params {
    tableNames: {
        schema: unknown;
        name: unknown;
    }[];
}
export interface Result {
    type: \\"Table\\";
    name: {
        schema: string;
        name: string;
    };
    columns: {
        name: string;
        isNullable: string;
        enum: string;
        type: string;
    }[];
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;

exports[`Traverse Should work: __generated__/tariff-rates.queries.ts 1`] = `
"export interface Params {
}
export interface Result {
}
export interface Query {
    params: Params;
    result: Result;
}
"
`;
